// b-link/lib/const.js
/**
 * @file 消息相关的常量
 * @author cxl(chenxinle@baidu.com)
 */

/**
 * 消息头部大小
 *
 * @const
 * @type {number}
 */
exports.HEADER_SIZE = 40;

/**
 * 设备标示
 *
 * @const
 * @type {Object}
 */
exports.DEVICE = {
    PC: 0,
    MOBILE: 1,
    MAC: 2 // TODO 待确认...
};

/**
 * 角色标示
 *
 * @const
 * @type {Object}
 */
exports.ROLE = {
    CSR: 0,
    VIS: 1
};

/**
 * 访客状态标示
 *
 * @const
 * @param {Object}
 */
exports.V_STATUS = {
    LOCKING: -1,    // Define by this module only, not blink
    ENTER: 0,
    ROUTING: 1,
    QUEUEING: 2,
    ALERTING: 3,
    TALKING: 4,
    LEAVE: 5
};

/**
 * 消息类型
 *
 * @const
 * @type {Object}
 */
exports.MSG = {
    NONE: 0,
    REQUEST: 1,
    RESPONSE: 2,
    NOTIFY: 3,
    FORWARD: 4,
    OFFLINE: 5,
    ERROR: 6,
    TIMEOUT: 7,
    ALL: 8
};

/**
 * 命令类型
 *
 * @const
 * @type {Object}
 */
exports.CMD = {
    VST_ENTER: 1,
    VST_LEAVE: 2,

    AGT_LOGIN: 7,
    AGT_LOGOUT: 8,
    AGT_CHANGESTATUS: 9,
    AGT_HEARTBEAT: 13,

    NTF_VSTAT: 15,
    NTF_CSTAT: 16,
    
    NTF_KICKOUT: 17,
    INVITE: 18,
    CONTENT: 19,
    ACCEPT: 20,
    REJECT: 21,
    BYE: 28,

    AGT_GET_COMLIST: 32,
    AGT_GET_COMSTAT: 33,
    PICK_QUEUE: 37,

    EX_HEARTBEAT: 1000,
    EX_ACK_MSG: 1002
};

/**
 * 在线状态
 *
 * @const
 * @type {Object}
 */
exports.STATUS = {
    ONLINE: 0,
    OFFLINE: 3,
    KICKOUT: 99
};

/**
 * 发送消息的类型
 *
 * @const
 * @type {Object}
 */
exports.CONTENT_TYPE = {
    HTML: 0,
    FILE: 1,
    VOICE: 2,
    VIDEO: 3,
    OTHER: 4
};

/**
 * 发送消息错误
 *
 * @const
 * @type {Object}
 */
exports.ERROR_SEND_MSG = {
    UNKNOW: 0, // 发送对象错误
    TIMEOUT: 1, // 发送消息超时
    DISCARD: 2  // 丢弃
};

/**
 * 请求结果类型
 *
 * @const
 * @type {Object}
 */
exports.RESULT = {
    SUCCESS: 0
};
