// b-link/lib/biz/bye.js
/**
 * @file 客服主动会话结束
 * @author mengran(mengran@baidu.com)
 */

var CONST = require('../const');
var getUserInfo = require('../util/getUserInfo');
var AckError = require('./AckError');
var Long = require('long');
var curry = require('../util/curry');

/**
 * Bye response
 *
 * @param {Object} res Response object
 */
function byeResponseHandler(resolve, res) {
    resolve(res);
}

/**
 * Bye error
 *
 * @param {Object} err Response object
 */
function byeErrorHandler(reject, err) {
    reject(err);
}

/**
 * Send bye request
 *
 * @public
 * @param {Object} core 沟通内核
 * @param {string} id identify of visitor
 * @return {Promise}
 */
module.exports = function (core, id) {
    var config = core._config;
    var contacts = core._getContacts(id);

    if (!contacts) {
        var error = new AckError('bye', {result: id});
        return Promise.reject(error);
    }

    var data = {
        from: core._getUserInfo(),
        fromtype: config.role,
        to: getUserInfo(contacts),
        totype: contacts.type,
        siteid: contacts.siteId,
        sessionid: contacts.sessionId,
        eid: config.eid,
        content: []
    };

    return new Promise(function (resolve, reject) {
        core._attachMsg(
            CONST.MSG.RESPONSE,
            CONST.CMD.BYE,
            curry(byeResponseHandler, resolve),
            true   // Once
        );
        core._attachMsg(
            CONST.MSG.ERROR,
            CONST.CMD.BYE,
            curry(byeErrorHandler, reject),
            true    // Once
        );
        core._send(CONST.MSG.REQUEST, CONST.CMD.BYE, data);
    });
};
