// b-link/lib/biz/changeStatus.js
/**
 * @file 客服改变状态
 * @author mengran(mengran@baidu.com)
 */

var CONST = require('../const');
var getUserInfo = require('../util/getUserInfo');
var AckError = require('./AckError');
var Long = require('long');
var curry = require('../util/curry');

/**
 * response
 *
 * @param {Object} res Response object
 */
function changeStatusResponseHandler(core, resolve, reject, status, res) {
    var body = res.body;
    var data = {
        before: body.status,
        after: status
    };
    
    if (body.result == CONST.RESULT.SUCCESS) {
        core._status = status;
        core._config.status = status;
        resolve(data);
    } else {
        reject();
    }
}

/**
 * 客服改变状态协议处理
 *
 * @public
 * @param {Object} core 沟通内核
 * @param {Number} status 改变的状态
 * @return {Promise}
 */
module.exports = function (core, status) {

    var currentUser = core._getUserInfo();
    var userInfo = core.getUserInfo();
    var data = {
        user: currentUser,
        eid: userInfo.eid,
        device: CONST.DEVICE.PC,
        token: userInfo.token,
        status: status
    };

    return new Promise(function (resolve, reject) {
        if (core._status != status) {
            core._attachMsg(
                CONST.MSG.RESPONSE,
                CONST.CMD.AGT_CHANGESTATUS,
                curry(changeStatusResponseHandler, core, resolve, reject, status),
                true    // Once
            );
            core._send(CONST.MSG.REQUEST, CONST.CMD.AGT_CHANGESTATUS, data);
        } else {
            reject();
        }
    });
};
