// b-link/lib/biz/getContactsList.js
/**
 * @file 获取联系人列表
 * @author cxl(chenxinle@baidu.com)
 */

var CONST = require('../const');
var curry = require('../util/curry');
var AckError = require('./AckError');

/**
 * 用户列表响应数据处理
 *
 * @param {Function} resolve resolve
 * @param {Function} reject reject
 * @param {Object} msg 消息对象
 */
function handler(resolve, reject, msg) {
    var core = this;
    var data = msg.body;

    if (data.result !== CONST.RESULT.SUCCESS) {
        reject(new AckError('getcontacts', data));
        return;
    }

    // 只处理访客
    // 暂时忽略客服
    var contactsList = [];
    data.visitors.forEach(function (item) {
        var contacts = {
            id: item.user.id,
            authType: item.user.authtype,
            nickname: item.nickname,
            eid: item.eid,
            device: item.device,
            siteId: item.siteid,
            type: CONST.ROLE.VIS
        };

        core._addContacts(contacts);
        contactsList.push(contacts);
    });

    resolve(contactsList);
}

/**
 * 获取用户列表
 *
 * @public
 * @param {Object} core 沟通内核
 * @return {Promise}
 */
module.exports = function (core) {
    var config = core._config;

    var data = {
        user: core._getUserInfo(),
        eid: config.eid,
        token: config.token
    };

    return new Promise(function (resolve, reject) {
        core._send(CONST.MSG.REQUEST, CONST.CMD.AGT_GET_COMLIST, data);
        core._attachMsg(
            CONST.MSG.RESPONSE,
            CONST.CMD.AGT_GET_COMLIST,
            curry(handler, resolve, reject),
            true
        );
    });
};
